hw06-atm/src/main/java/ru/otus/javadeveloper/hw06/atm/core/atm/ActiveRecoverAtm.java

```diff
@@ -0,0 +1,49 @@
+package ru.otus.javadeveloper.hw06.atm.core.atm;
+
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import ru.otus.javadeveloper.hw06.atm.BankNote;
+import ru.otus.javadeveloper.hw06.atm.core.command.Command;
+import ru.otus.javadeveloper.hw06.atm.core.command.WriteOffWithAutoRecover;
+import ru.otus.javadeveloper.hw06.atm.core.container.BankNoteContainer;
+import ru.otus.javadeveloper.hw06.atm.core.container.BankNoteContainerResponse;
+import ru.otus.javadeveloper.hw06.exceptions.AtmHasNotEnoughBanknotesException;
+
+import java.util.List;
+
+@Slf4j
+@NoArgsConstructor
+public class ActiveRecoverAtm implements FullFunctionalAtm {
```

Comment:
Вы вроде разделили интерфейсы на BaseAtm и CheckableAndRecoverable в соответствии с  Interface Segregation Principle, но потом объединили их в FullFunctionalAtm.

Если придерживаться разделения интерфейса, то должно быть:
```suggestion
public class ActiveRecoverAtm implements BaseAtm, CheckableAndRecoverable  {
```

И дальше работать с BaseAtm, CheckableAndRecoverable 

А точнее даже так:
```suggestion
public class ActiveRecoverAtm implements BaseAtm, CheckableBalance {
```
Это во вопросу "Как избежать компромисса фиктивной реализации метода recover у AutoRecoverAtm?"

Reply:
Да, но в Department используется CheckableAndRecoverable, удалив этот интерфейс придется создавать параллельно 2 списка CheckableBalance и ManuallyRecoverable и для каждого из них метод add/remove, а при добавлении в департамент объекты типов LazyRecoverAtm или NotRecoverableAtm придется добавлять их дважды, поэтому и пришлось придумывать искуственный CheckableAndRecoverable.

С CheckableAndRecoverable похожая история, но уже в тестах, где нужно одновременно использовать и writeOff и getBalance. Единственным выходом вижу кастить руками при каждом выполнении одной из двух операций.

Reply:
Гугление открыло еще одно решение, по-крайней мере для замены CheckableAndRecoverable интерфейса, можно изменить addObserver(Object, Class), таким образом внутри Department через рефлексию добавлять либо в один список, либо в оба, в зависимости от того какой набор интерфейсов реализует класс. Но хорошее ли это решение через рефлексию?

Reply:
Рефлексия как вариант, но не красиво в этом ДЗ.

Если просто убрать CheckableAndRecoverable, то такой вариант:
```java
public class Department implements CheckableBalance, ManuallyRecoverable {
    private List<Object> atmList = new ArrayList<>();

    public <T extends CheckableBalance & ManuallyRecoverable> void removeObserver(T atm){
        atmList.remove(atm);
    }

    public <T extends CheckableBalance & ManuallyRecoverable> void addObserver(T atm){
        atmList.add(atm);
    }

    @Override
    public long getBalance() {
        return atmList.stream().mapToLong(atm -> ((CheckableBalance) atm).getBalance()).sum();
    }

    @Override
    public void recover() {
        atmList.forEach(atm -> ((ManuallyRecoverable) atm).recover());
    }
}
```

Или подумать над архитектурой в целом.

По поводу улучшения `List<Object> atmList` - [здесь](https://stackoverflow.com/questions/27913542/is-it-possible-to-define-a-list-of-any-type-that-implements-multiple-interfaces) посмотрите